src/main/java/org/jboss/netty/channel/socket/nio/AbstractNioChannel.java

```diff
@@ -256,10 +257,14 @@ public boolean offer(MessageEvent e) {
             if (newWriteBufferSize >= highWaterMark) {
                 if (newWriteBufferSize - messageSize < highWaterMark) {
                     highWaterMarkCounter.incrementAndGet();
-                    if (!notifying.get() && setUnwritable()) {
-                        notifying.set(Boolean.TRUE);
-                        fireChannelInterestChanged(AbstractNioChannel.this);
-                        notifying.set(Boolean.FALSE);
+                    if (setUnwritable()) {
+                        if (!isIoThread(AbstractNioChannel.this)) {
+                            fireChannelInterestChangedLater(AbstractNioChannel.this);
+                        } else if (!notifying.get()) {
```

Comment:
We could check `notifying.get()` first to avoid `isIoThread()` call completely:

```
if (!notifying.get()) {
    if (!isIoThread(AbstractNioChannel.this)) { ... } else { ... }
}
```

Same applies to the change below.